amino.py

```diff
@@ -176,55 +202,57 @@ def add_OP(self, OP):
-            update = False
-            difference = None
-            for i in range(len(self.OPs)):
-                candidate_info = 1
-                for j in range(len(self.OPs)):
-                    if not i == j:
-                        candidate_info = candidate_info * mut_info[j]
-                if candidate_info > existing[i]:
-                    update = True
-                    if difference == None:
-                        difference = candidate_info - existing[i]
-                        old_OP = i
-                    else:
-                        if (candidate_info - existing[i]) > difference:
-                            difference = candidate_info - existing[i]
-                            old_OP = i
-            if update == True:
-                mut_info[old_OP] = self.mut.distance(OP, OP)
-                self.matrix[old_OP] = mut_info
-                self.OPs[old_OP] = OP
-                for i in range(len(self.OPs)):
-                    self.matrix[i][old_OP] = mut_info[i]
-        else:
-            for i in range(len(self.OPs)):
-                mut_info = self.mut.distance(OP, self.OPs[i])
-                self.matrix[i].append(mut_info)
-                self.matrix[len(self.OPs)].append(mut_info)
-            self.matrix[len(self.OPs)].append(self.mut.distance(OP, OP))
+        if len(self.OPs) == self.size: # matrix is full, check for swaps
+
+            # distance of the new OP to all existing OPs
+            mut_info = np.array([self.mut.distance(OP, i) for i in self.OPs])
+
+            # product of distances between all existing OPs, excluding the diagonal element (Eqn. 3)
+            # Taking the power of 1/(|S|-1) is not necessary for comparison purposes
+            existing = np.array([np.prod(self.matrix[i], where=(np.arange(self.size) != i)) for i in np.arange(self.size)])
+
+            for i in np.arange(self.size):
+
+                n = np.argmin(existing)   # Algorithm 3 Line 7
+
+                candidate_info = np.prod(mut_info, where=(np.arange(self.size) != i)) # Algorithm 3 Line 8
+
+                if existing[n] < candidate_info:  # Algorithm 3 Line 9
+
+                    mut_info[n] = self.mut.distance(OP, OP) 
+                    self.matrix[n,:] = mut_info
+                    self.matrix[:,n] = mut_info
+                    self.OPs[n] = OP  # Update the list of OPs
+
+        else: # adding an OP when there are fewer than self.size
+
+            mut_info = np.zeros(self.size)
+
+            n = len(self.OPs)
+            for i, op in enumerate(self.OPs):
+                mut_info[i] = self.mut.distance(OP, op)
+            mut_info[n] = self.mut.distance(OP, OP)
+
+            try:
+                assert(mut_info[n] < 1e-2)
+            except:
+                raise ValueError("The dissimilarity between an OP and itself is not close to zero. Decrease bandwidth.")
+
+            self.matrix[n,:] = mut_info
+            self.matrix[:,n] = mut_info
```

Comment:
I think we can actually clean this up further since there is a lot of repeated functionality like adding `mut_info` to `matrix` and the `self.mut.distance` loop.

Maybe we should also write a unit test for this before we change it since it's becoming a more complicated refactor.